### PR TITLE
Add options to toggle QtCustomTitleBar

### PR DESCRIPTION
**Pull request by DannyAlas**
# Description

Adds the option to not render the `hide_btn` and `float_btn` in the QtCustomTitleBar of any widget. Previously this was only possible for the `close_btn`

Users can now toggle the rendering of `hide_btn` and `float_btn` in QtCustomTitleBar of any widget. For example: 
``` Python
from qtpy import QtWidgets
import napari

viewer = napari.Viewer()
myWidget = QtWidgets.QWidget()
w1 = viewer.window.add_dock_widget(myWidget)

# toggle by setting the respective button to True or False
w1._close_btn = False
w1._hide_btn = True
w1._float_btn = False
```
#4027 added support for the `close_btn` and the feature was also mentioned on [image.sc](https://forum.image.sc/t/can-i-remove-the-close-icon-when-i-create-a-dock-widget-in-the-viewer-with-add-dock-widget/67369/4)

## Type of change
normal
all
`_close_btn` hidden
Screenshot 2022-12-13 194325
`_close_btn`  and `_hide_btn` hidden
Screenshot 2022-12-13 194354
`_close_btn`,`_hide_btn`, and `_float_btn` hidden
Screenshot 2022-12-13 194423

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [x] all tests pass

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by brisvag** [comment]
Thanks for the PR @DannyAlas! I udnerstand the need to prevent closing the widged; but what is the use case for preventing hiding and undocking? I think these things would hust get annoying if plugins started deeming their widgets "too important" to get moved around :P

Depending on the usefulness, we might even expose something like this publicly, rather than with a private attribute.

**Comment by DannyAlas** [comment]
The purpose is really just to provide further control over custom widgets. In our use case, we're trying to allow the docking of widgets created in napari within our Qt app (e.g. those created by napari plugins).

In doing this, we decided to simply place the napari widget inside a QDockWidget created for our app if we detect an attempt to dock the widget within our app. If the user decides to place the widget back inside napari. We simply pass the widget back to napari and destroy the one created in our app. (this is a workaround so we don't need to design an entire backend for handling napari widgets, though we might endeavor on this in the future)

Ideally, we would like to disable/hide the custom titlebar or at least the rendering of the buttons to prevent them from appearing as below.
image

I would highly support a push to publicly manipulate the custom titlebar rather than with a private attributes. ~~However, I'd have to familiarize myself further with the codebase before I'd feel comfortable creating a pull request for this. For now, the bare minimum utility of at least being able to disable the button rendering seems quite useful.~~ I'll see if I can at least create a public-facing flag for disabling the custom title bar when creating a widget to get this started.
 Thanks!

**Comment by DannyAlas** [comment]
As mentioned above, I've attempted to add a public-facing flag for users to choose if they'd like to add the custom_title_bar when creating a docked widget using the `add_custom_title_bar` in `add_dock_widget`. 
For example:
```python
viewer = napari.Viewer(show=False)
widget = QWidget()
dock_widget = viewer.window.add_dock_widget(widget, add_custom_title_bar=False)
```
would render a dock widget that looks like this
image

Take this use case: 
I'd like to manually dock a napari dock_widget inside a QApplication with the inbuilt hooks and signals napari already integrates for a dock_widget. 

```python
class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        # set the main application window
        self._main = QWidget()
        self.setCentralWidget(self._main)
        self.layout = QGridLayout(self._main)
        
        # create napari viewer
        self.viewer = napari.Viewer(show=False)

        ##############################################
        # this is a workaround for napari issue #5400
        # initialize qt_viewer and qt_window
        self._qt_viewer = self.viewer.window._qt_viewer

        self.qt_window = self.viewer.window._qt_window
        ##############################################

        self.layout.addWidget(self.qt_window, 0, 0)
        
        self.init_dock_widget()
    
    # create a magicgui widget that has layer selections
    @magicgui(image = {'label': 'Pick an Image'})
    def example_func(image: Image):
        # do something with image
        return image

    def init_dock_widget(self):
        # add the magicgui widget to the napari 
        # viewer as a dock widget
        napari_widget = self.viewer.window.add_dock_widget(
            self.example_func, 
            add_custom_title_bar=False,
            )

        # mannually add the napari dock widget to our main 
        # window as a QDockWidget
        self.addDockWidget(Qt.RightDockWidgetArea, napari_widget)
```
Using the default value of `add_custom_title_bar=True` when adding a dock widget yields the following result: no button icons and the custom title bar is hard to differentiate from the widget.
image

changing `add_custom_title_bar=True` to `add_custom_title_bar=False` in the napari widget creation now yields 
image

I believe this allows for better implementation of napari widgets within other Qt applications and gives the user more control over dock widgets. I'm quite the novice when it comes to python, so feel free to let me know if I've completely missed something. Thanks!!

**Comment by brisvag** [comment]
Interesting! I think @Czaki might be better suited to help/comment on this, since he's experienced with embedding napari in another qt app.

In my opinion, it's fine if we add this option to `add_dock_widget()`, but would advise against exposing it to the plugin interface to prevent abuse :)

**Review by DannyAlas — COMMENTED** [review]
typo
the commit should be: Fix formatting

File: napari/_qt/widgets/qt_viewer_dock_widget.py [hunk 1]
```diff
@@ -316,19 +337,24 @@ def __init__(
...
+        if float_btn:
+            self.float_button = QPushButton(self)
+            self.float_button.setToolTip(trans._('float this panel'))
+            self.float_button.setObjectName("QTitleBarFloatButton")
+            self.float_button.setCursor(Qt.CursorShape.ArrowCursor)
+            self.float_button.clicked.connect(
+                lambda: self.parent().setFloating(
+                    not self.parent().isFloating()
+                )
+            )
```
Comment (Czaki) [inline napari/_qt/widgets/qt_viewer_dock_widget.py]: Hm. I prefer to control visibility, not the presence of buttons. Control presence breaks widget API.

**Review by Czaki — COMMENTED** [review]
Hm. 
There is few of things that I do not understand.

1) why do you not apply the napari style to the whole application? Now you have incompatible parts of the application if we think about style. 

2) Why did you dock the whole main window not `QtViewer`?

3) You would like to modify part of the private napari API. We assume that if we need, we could break it without backward compatibility. If you think that `QtViewerDockWidget` could be helpful for people creating applications based on napari, then it should be reexported in `napari.qt` module. 

4) If you want to remove these buttons but keep using `QtViewerDockWidget` then I understand that Your task is to allow users to rearrange widgets, but not allow them to close them?

**Comment by DannyAlas** [comment]
@Czaki let me know if this makes more sense!

> 1. why do you not apply the napari style to the whole application? Now you have incompatible parts of the application if we think about style.

We could apply the napari stylesheet to the application, but for the purpose of a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) for just removing the buttons, I didn't find it necessary. In the future, it may be ideal to apply custom non-napari stylesheets to match the theme of different programs that could integrate napari.

> 2. Why did you dock the whole main window not QtViewer?

Docking the entire application was just for the simplicity of having the other dock widgets instead of just the Qt viewer. But we have started to do everything as individual dock widgets, which entails either creating our own stylesheets or using the napari style sheets for everything. For the purpose of hiding the buttons, though, I don't think this makes a difference.

> 3. You would like to modify part of the private napari API. We assume that if we need, we could break it without backward compatibility. If you think that QtViewerDockWidget could be helpful for people creating applications based on napari, then it should be reexported in napari.qt module.

If I understand correctly, you're stating that this change should be exported to the [napari.qt](https://github.com/napari/napari/tree/main/napari/qt) module as a public-facing element for backward compatibility? If so, having a look at the other files there, I'm not too sure I understand the purpose/structure of the module, is there somewhere that explains this? Should these changes just be implemented as public-facing arguments passthrough to the private elements? 

> 4. If you want to remove these buttons but keep using QtViewerDockWidget then I understand that Your task is to allow users to rearrange widgets, but not allow them to close them?

No. The purpose of this is to allow for greater customization of the dock widgets created for napari. Our purpose of using napari in a larger application seems to be somewhat unique, but the feature is simply to allow dock widgets to have more control over the title bar. As it stands, this custom title bar is added forcefully to any widgets created in napari, and I believe having the ability to at least disable it could be nice (maybe even allowing users to create their own as well).

File: napari/_qt/widgets/qt_viewer_dock_widget.py (hunk 1 above)
Comment (DannyAlas) [reply napari/_qt/widgets/qt_viewer_dock_widget.py]: Would it be okay to instead control the presence of the custom title bar?

**Comment by Czaki** [comment]
> If I understand correctly, you're stating that this change should be exported to the [napari.qt](https://github.com/napari/napari/tree/main/napari/qt?rgh-link-date=2022-12-22T23%3A20%3A46Z) module as a public-facing element for backward compatibility

No, this is a package for QT elements that we agree to expose so that we will promise to trey to keep a more stable interface. We keep all implementations in `napari._qt` to avoid circular import. 

> No. The purpose of this is to allow for greater customization of the dock widgets created for napari. Our purpose of using napari in a larger application seems to be somewhat unique

I don't think. Would Your application be much larger than PartSeg? I also see other screenshots of larger applications that embed napari. This part of usage is not documented but is already present. 

> No. The purpose of this is to allow for greater customization of the dock widgets created for napari

The whole idea around dock widgets is to allow users to arrange widgets on his own. It looks like you don't like it, so why would you use a dock widget instead of a plain `QWidget`? 

It is not that I'm against this change. But I know the napari codebase quite well (especially qt part), and it looks like you are trying to do something in a nonoptimal way that may hurt in the future. Because of this, I'm just trying to ask what you would like to achieve, not how you would like to achieve this.

**Comment by DannyAlas** [comment]
> The whole idea around dock widgets is to allow users to arrange widgets on his own. It looks like you dont like it, so why would you use a dock widget instead of a plain QWidget?

I understand. It's not that I don't like napari's dock widgets, I actually really like them, I would just like to have the **option to disable the custom title bar**. As far as I can tell, this doesn't affect the ability to move the widgets, and the stock title bar still has the float and close buttons, see below
image

In the future, it may also be nice for users to have the option to create their own custom title bars for dock widgets though that's outside of the scope of this pull request.

---

As per implementation, this may not be the optimal way to do this feature, so any advice is welcome. I'm very inexperienced with napari and Qt in general, so feel free to critique away! I also don't think our application is necessarily larger than PartSeg, just different.

**Review by alisterburt — COMMENTED** [review]
Hi @DannyAlas, super cool to hear you're integrating napari into your own application, really exciting!

I'm not against adding this customisability to existing dock widgets/the `add_dock_widget()` method - it seems reasonable and as @brisvag says, as long as it isn't exposed via the plugin interface I don't see it as problematic. @Czaki is there a specifc reason you are worried or is it just that you do not understand the motivation? 

I think without knowing more about what exactly @DannyAlas is trying to do in his application we won't understand the motivation. I would love to know the envisioned use case in more detail but even without that I find the requested change reasonable

I'm not sure I understand the implementation though, it looks like the hide button and float button have been removed and I don't see where else they are readded?

**Comment by Czaki** [comment]
In the current state of this PR some plugin may add a widget that cannot be moved or removed by a user. 
This change is not limited to embedding napari in another application, but in every scenario, that may significantly impact inexperienced users.

**Comment by DannyAlas** [comment]
Hi @alisterburt! Thank you so much! Our use case isn't even fully fleshed out, and currently, we're just trying to implement methods for manual ROI manipulation, though when we get something working, I'd love to share if there's a place for that.

Here is an example of this PR in action where the widget is created using
```python
import napari 
from qtpy.QtWidgets import QWidget

viewer = napari.Viewer()
widget = QWidget()

viewer.window.add_dock_widget(widget, add_custom_title_bar=False, menu=viewer.window.window_menu)
```
example of dock_widget with no titlebar

@Czaki, I might be misunderstanding something, but as far as I can tell, this PR would still allow users to close and move the widgets, as seen above. Does this PR change something in the plugin API somewhere I'm not noticing (I haven't used it much)?

Thanks, guys!

**Comment by GenevieveBuckley** [comment]
I'm also having trouble clearly understanding your motivation.
Yes, we do like minimal examples for implementation demos (the "what", or behaviour of a new change). But to fully understand your motivation (the "why"), knowing the full context & describing a concrete example is very helpful.

Can you explain it using a framework like this?
1. **The broader context**
    * It sounds like you want to stick napari with some dock widgets inside your own Qt application. This application does something to do with manual ROI manipulation.
2. **Your specific aim, and why**
    * You say "I would just like to have the option to disable the custom title bar". Can you explain why you need/want this for your broader goal?
3. **Your current solution/implementation (before this PR)**
    * It sounds like you've embedded napari into your own application, and are using a lot of individual napari dock widgets, and creating your own custom style sheets for this. Is that correct?
4. **Why you don't like this current situation, what is the problem this is currently causing?**
    * This is what I don't understand. You don't want the float and hide buttons displayed on the docked widget(s)... but you don't actually care whether the user can move or hide the widget(s)? Why? Is it just because you want it to "look nicer" in some way, and/or because you don't want to use style sheets? What is the current problem?
5. **Suggested change, and how this fixes the problem**
    * The suggested change is this PR, but because I don't understand what the problem is, it's hard to see if this fixes it! (It also makes it hard to judge whether it's likely to be useful for many other users or not, which changes how likely it is for the PR to be accepted)

File: napari/_qt/widgets/qt_viewer_dock_widget.py [hunk 2]
```diff
@@ -352,12 +350,6 @@ def __init__(
                 layout.addWidget(
                     self.close_button, 0, Qt.AlignmentFlag.AlignHCenter
                 )
-            layout.addWidget(
-                self.hide_button, 0, Qt.AlignmentFlag.AlignHCenter
-            )
-            layout.addWidget(
-                self.float_button, 0, Qt.AlignmentFlag.AlignHCenter
-            )
```
Comment (GenevieveBuckley) [inline napari/_qt/widgets/qt_viewer_dock_widget.py]: I will also say that if we go ahead with the idea of allowing users to provide their own custom title bars, I think it's better if people define their own Qt custom title bar class in their own code, and pass that in to napari somehow.

Just deleting some bits out of the napari codebase to suit one specific use case is probably not going to work well as soon as the next group or several come along with their own conflicting ideas about how they want to customize title bars a slightly different way.

File: napari/_qt/qt_main_window.py [hunk 3]
```diff
@@ -863,6 +863,7 @@ def add_dock_widget(
         allowed_areas: Optional[Sequence[str]] = None,
         shortcut=_sentinel,
         add_vertical_stretch=True,
+        add_custom_title_bar=True,
```
Comment (Czaki) [inline napari/_qt/qt_main_window.py]:
```suggestion
        use_napari_title_bar=True,
```

File: napari/_qt/qt_main_window.py [hunk 4]
```diff
@@ -896,6 +897,8 @@ def add_dock_widget(
                 The shortcut parameter is deprecated since version 0.4.8, please use
                 the action and shortcut manager APIs. The new action manager and
                 shortcut API allow user configuration and localisation.
+        add_custom_title_bar : bool, optional
```
Comment (Czaki) [inline napari/_qt/qt_main_window.py]:
```suggestion
        use_napari_title_bar : bool, optional
```

File: napari/_qt/qt_main_window.py [hunk 5]
```diff
@@ -930,6 +933,7 @@ def add_dock_widget(
                 allowed_areas=allowed_areas,
                 shortcut=shortcut,
                 add_vertical_stretch=add_vertical_stretch,
+                add_custom_title_bar=add_custom_title_bar,
```
Comment (Czaki) [inline napari/_qt/qt_main_window.py]:
```suggestion
                use_napari_title_bar=use_napari_title_bar,
```

File: napari/_qt/qt_main_window.py [hunk 6]
```diff
@@ -939,6 +943,7 @@ def add_dock_widget(
                 area=area,
                 allowed_areas=allowed_areas,
                 add_vertical_stretch=add_vertical_stretch,
+                add_custom_title_bar=add_custom_title_bar,
```
Comment (Czaki) [inline napari/_qt/qt_main_window.py]:
```suggestion
                use_napari_title_bar=use_napari_title_bar,
```

File: napari/_qt/widgets/qt_viewer_dock_widget.py [hunk 7]
```diff
@@ -64,6 +64,9 @@ class QtViewerDockWidget(QDockWidget):
         Whether to add stretch to the bottom of vertical widgets (pushing
         widgets up towards the top of the allotted area, instead of letting
         them distribute across the vertical space).  By default, True.
+    add_custom_title_bar : bool, optional
```
Comment (Czaki) [inline napari/_qt/widgets/qt_viewer_dock_widget.py]:
```suggestion
    use_napari_title_bar : bool, optional
```

File: napari/_qt/widgets/qt_viewer_dock_widget.py [hunk 8]
```diff
@@ -77,12 +80,14 @@ def __init__(
         shortcut=_sentinel,
         object_name: str = '',
         add_vertical_stretch=True,
+        add_custom_title_bar=True,
```
Comment (Czaki) [inline napari/_qt/widgets/qt_viewer_dock_widget.py]:
```suggestion
        use_napari_title_bar=True,
```

File: napari/_qt/widgets/qt_viewer_dock_widget.py [hunk 9]
```diff
@@ -77,12 +80,14 @@ def __init__(
...
         close_btn=True,
     ):
         self._ref_qt_viewer: 'ReferenceType[QtViewer]' = ref(qt_viewer)
         super().__init__(name)
         self._parent = qt_viewer
         self.name = name
+        self.add_custom_title_bar = add_custom_title_bar
```
Comment (Czaki) [inline napari/_qt/widgets/qt_viewer_dock_widget.py]:
```suggestion
        self.use_napari_title_bar = use_napari_title_bar
```

File: napari/_qt/widgets/qt_viewer_dock_widget.py [hunk 10]
```diff
@@ -146,20 +151,22 @@ def __init__(
...
-        )
-        self.setTitleBarWidget(self.title)
+        # add custom title bar
+        if add_custom_title_bar:
+            self.title = QtCustomTitleBar(
+                self,
+                title=self.name,
+                close_btn=close_btn,
+            )
+            self.setTitleBarWidget(self.title)
```
Comment (Czaki) [inline napari/_qt/widgets/qt_viewer_dock_widget.py]:
```suggestion
        if use_napari_title_bar:
            self.title = QtCustomTitleBar(
                self,
                title=self.name,
                close_btn=close_btn,
            )
            self.setTitleBarWidget(self.title)
        else:
            self.title = None
```

File: napari/_qt/widgets/qt_viewer_dock_widget.py [hunk 11]
```diff
@@ -305,7 +312,11 @@ class QtCustomTitleBar(QLabel):
     """
 
     def __init__(
-        self, parent, title: str = '', vertical=False, close_btn=True
+        self,
+        parent,
+        title: str = '',
+        vertical=False,
+        close_btn=True,
```
Comment (Czaki) [inline napari/_qt/widgets/qt_viewer_dock_widget.py]: Why are changes in this file required? I think that it should be reverted.

**Review by Czaki — COMMENTED** [review]
After a long break, I rethink my objections to this PR.  

First, I would like to rename.`add_custom_title_bar` to use `use_napari_title_bar` or something similar. The custom title bar refers to implementation details, not this parameter's semantic meaning and the source of my confusion. 

When I see `add_custom_title_bar`, I expect to pass the Qt class implementing the title bar, not the bool. 

Even with this change, this PR does not fit motivation. 
Everything that lives in `napari._qt` is not exported to `napari.qt` is part of the private napari API, and could be changed without warning. 

So with code from the current state of PR, we could totally reimplement dockable widgets with only satisfy API of `viewer.window.add_dock_widget`. But you motivate PR by being able to dock napari plugins somewhere in your application. So, to have this working in a stable way, there is a need to make docking-related class part of public API.

File: napari/_qt/widgets/qt_viewer_dock_widget.py (hunk 11 above)
Comment (DannyAlas) [reply napari/_qt/widgets/qt_viewer_dock_widget.py]: I believe lines 314 to 319 were just my linter. The change isn't required and can be reverted.

**Comment by DannyAlas** [comment]
Hi @GenevieveBuckley! I'll attempt to explain my original motivations using the framework you provided. I'm very new to Qt and GUI development in general, so please let me know if I'm missing something.

**The broader context:**
- I have a Qt application that I would like to integrate Napari into to handle manual ROI manipulation. As this was only a small part of a larger feature, I was hoping to somehow dock the entire Napari application with its dock widgets inside my app. This is as opposed to what I've now done, which is to write a proper API to handle and dock the individual widgets in my app.

**Specific aim:**
- While the Napari application works fine inside another Qt app, the dock widgets have a custom title bar that would require special handling and styling if placed in another Qt app, as seen in [5407#issuecomment-1352524956](https://github.com/napari/napari/pull/5407#issuecomment-1352524956 ), the rest of the dock widget still works fine, including all the event handlers. **The aim of this PR** was to fix this by allowing users to choose if they would like to use the `QtCustomTitleBar` or stick with the standard Qt `TitleBarWidget`.

**What is the problem:**
- The QtCustomTitleBar is forcefully added to a dock widget without a public method to manipulate it, remove it or pass a custom one. If a user wishes to dock these widgets in another app, they'll end up with the [aforementioned issue](https://github.com/napari/napari/pull/5407#issuecomment-1352524956 ), and if a user wishes to just add or manipulate the title bar of a widget, there is no method to do so without accessing the private `_qt_viewer` class.

**Suggested change:**
- I thought the simplest method to fix these issues was to give the user the option to use the `QtCustomTitleBar` or the standard Qt TitleBarWidget within the `viewer.window.add_dock_widget` method.

With that said, though, as @Czaki [mentioned ](https://github.com/napari/napari/pull/5407#pullrequestreview-1243645220) this issue seems to be hitting a more fundamental problem, which is that 
> there is a need to make docking-related class part of public API

I believe this feature could be better handled by a new PR for **Creating a public API for docking widgets** which allows users to pass in a custom title bar. For that reason, I will close this PR and move to work on that implementation.

**Pull request closed without merging**